### DISCUSSION TITLE
Issue with API Access When Running GHOSTS SHADOWS via Docker

### DISCUSSION THREAD

**Discussion by dh4340** [cmu-sei/GHOSTS#406] [Q&A]
Hi,

I'm having trouble accessing the API for GHOSTS SHADOWS when running it via Docker. The setup works perfectly on bare metal, but when I try to use Docker, I encounter issues.

### Context:
- **GHOSTS SHADOWS** provides several interfaces, including a REST API and a UI web interface.
- **Endpoints** include activities, chat, content, and social.
- **Bare Metal Setup**:
  - `ollama create` and `ollama run` commands are executed for various models.
  - `python api.py` and `python ui.py` are run in separate terminals.
  - `OLLAMA_HOST=0.0.0.0:11434 ollama serve` is used to make the API available beyond localhost.

- **Docker Setup**:
  - Using the following command to run GHOSTS SHADOWS:
    ```bash
    EXPORT GHOSTS_OLLAMA_URL=http://localhost:11434
    docker run -d --name shadows -p 5900:5900 -p 7860:7860 dustinupdyke/ghosts-shadows
    ```
  - The container is started but I cannot call the API endpoints. For example, attempting to `curl` the chat endpoint results in a connection error.

### Error Details:
- When attempting to call the API via Docker, I receive a `ConnectionRefusedError` indicating that the service is not accessible at `http://localhost:11434/api/chat`.

### Things I Have Tried:
- Verified that the container is running and listening on the correct ports.
- Checked Docker network configurations to ensure that the container can communicate as expected.
- Tried accessing the API from both within the Docker container and from the host.

### Question:
Has anyone encountered a similar issue with Docker and GHOSTS SHADOWS? Are there specific Docker configurations or environment variables that might be affecting connectivity? Any suggestions or troubleshooting steps would be greatly appreciated.

Thank you!

**Comment by sei-dupdyke** [comment]
The tricky thing about containers is that localhost inside the container stays inside the container, so to reach localhost on the host from inside the container, use something like your host ip address. This is a change to this part of the command:

`EXPORT GHOSTS_OLLAMA_URL=http://<your_ip_address>:11434`

Let me know if that fixes the issue, and thanks for checking out the project.

**Comment by dh4340** [comment]
Didn't work, currently I have it running with just the API up and not the UI locally as can't figure out how to get it to correctly expose as per above.

**Comment by sei-dupdyke** [comment]
Are you running ollama on your host machine or also in a container?

**Reply by dh4340** [reply]
i was running ollama on host, then tried in container still no luck.
